src/openms/source/METADATA/SpectrumLookup.cpp

```diff
@@ -0,0 +1,276 @@
+// Copyright The OpenMS Team -- Eberhard Karls University Tuebingen,
+// ETH Zurich, and Freie Universitaet Berlin 2002-2015.
+//
+// This software is released under a three-clause BSD license:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Hendrik Weisser $
+// $Authors: Hendrik Weisser $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/METADATA/SpectrumLookup.h>
+
+using namespace std;
+
+namespace OpenMS
+{
+  const String& SpectrumLookup::default_scan_regexp = "=(?<SCAN>\\d+)$";
+
+  const String& SpectrumLookup::regexp_names_ = "INDEX0 INDEX1 SCAN ID RT";
+
+  SpectrumLookup::SpectrumLookup(): 
+    rt_tolerance(0.01), n_spectra_(0),
+    regexp_name_list_(ListUtils::create<String>(regexp_names_, ' '))
+  {}
+
+
+  SpectrumLookup::~SpectrumLookup()
+  {}
+
+
+  bool SpectrumLookup::empty() const
+  {
+    return n_spectra_ == 0;
+  }
+  
+
+  Size SpectrumLookup::findByRT(double rt) const
+  {
+    double upper_diff = numeric_limits<double>::infinity();
+    map<double, Size>::const_iterator upper = rts_.upper_bound(rt);
+    if (upper != rts_.end())
+    {
+      upper_diff = upper->first - rt;
+    }
+    double lower_diff = numeric_limits<double>::infinity();
+    map<double, Size>::const_iterator lower = upper;
+    if (lower != rts_.begin())
+    {
+      --lower;
+      lower_diff = rt - lower->first;
+    }
```

Comment:
Note that `lower` is initialized to `upper`; there's no call to `std::lower_bound`. So if `lower` (`== upper`) is at `rts_.begin()`, we can ignore `lower` and only consider `upper`. Then it's not necessary to calculate the difference again.